### ISSUE TITLE
Uploading: write to a writable stream, or pass in a readable stream?

### ISSUE THREAD

**Issue by domenic** [yutakahirano/fetch-with-streams#30]
Previously: https://github.com/yutakahirano/fetch-with-streams/pull/3 and https://github.com/yutakahirano/fetch-with-streams/issues/25#issuecomment-89826020

My opinion is that writing to a response body, i.e. doing an upload or writes to a cache, should be represented by a writable stream. It feels like a major failure of the entire writable stream design if we can't use it for cases like this.

It's true that in general you never "need" writable streams, as you can always have functions that accept readable streams and then directly feed their contents to underlying resources (e.g. directly call write(2) on the cache or send(2) on the socket). But kind of the entire point is to have an abstraction that lets you hide away that complexity.

@wanderview is not so sure...

It's also worth noting there is an equivalence here:

- Given `function f(bodyWriter) { bodyWriter(ws); }`, we can get `g(rs)` via
   ```js
   function g(rs) {
     f(ws => rs.pipeTo(ws));
   }
   ```

- Given `function g(rs) { doSomethingWith(rs); }`, we can get `f(bodyWriter)` via
  ```js
  function f(bodyWriter) {
    const ts = new TransformStream(/* identity transform */);
    bodyWriter(ts.writable);
    g(ts.readable);
  }
  ```

But I am more interested in what is natural to implementations. My hope would be that you could supply a writable stream that would directly wrap to the socket for upload, or file descriptor for write to cache, and not have to do any adaptation of the type shown in the second example.

**Comment by tyoshino** [comment]
Let's think of uploading chunks of raw bytes manually. I don't want to write code (create a ReadableStream and enqueue data to the ReadableStream) to pretend to be a ReadableStream. I'd rather want the Request object to expose some simple API using which I can write the chunks. That's the WritableByteStream API.

We also need a method to upload data from a ReadableStream. We can have a separate API surface that accepts a ReadableStream and consumes data to upload from it. But we can represent that as `rs.pipeTo(req.ws)`.

I agree with Ben about the point that a Request is not always wrapping a socket, etc.
- In the idea that a Request exposes a WritableStream, a Request is basically holds a buffer, and once it's passed to cache, etc. it starts also wrapping a socket, file, etc.
- In the idea that a Request doesn't expose a WritableStream but holds a reference to a ReadableStream, a Request is basically nothing but a ReadableStream reference holder, and the ReadableStream works as a buffer.

**Comment by tyoshino** [comment]
> I don't want to write code (create a ReadableStream and enqueue data to the ReadableStream) to pretend to be a ReadableStream

We could make this better by making the underlying source more API user friendly by using a Promise, and reduce the code to construct a ReadableStream (define an underlying source class and call new ReadableStream()). Then, that's the same as exposing the WritableStream on a Request.

**Comment by wanderview** [comment]
I also like the idea that passing a ReadableStream to the Constructor means we have exactly one place in the API for specifying the body a Request/Response.

I thought we already had a way to convert a push source to a ReadableStream.  Is that not adequate here for some reason?

**Comment by domenic** [comment]
> I also like the idea that passing a ReadableStream to the Constructor means we have exactly one place in the API for specifying the body a Request/Response.

I think that's solvable independently. It does seem like we eventually landed on overloading `body` such that when it's a function it acts as a body writer. My `bodyWriter` was an older idea, reviewing those threads.

> I thought we already had a way to convert a push source to a ReadableStream. Is that not adequate here for some reason?

Well, that mechanism is generally used for wrapping lower-level things, and not supposed to be used by normal authors, but instead by library writers who are creating custom streams to wrap "legacy" push sources. (Similar to library writers creating promises to wrap "legacy" callback code.)

As @tyoshino mentions, the underlying source API is not as well-suited as the writable stream API. For example when you enqueue something from an underlying source into a readable stream, that can never fail, so it just returns `undefined`. Whereas when you upload something to a server or write it to a cache, it can fail---that makes `ws.write(chunk)`, returning a promise, a better fit.

**Comment by wanderview** [comment]
@domenic Can you write some example code showing where passing a ReadableStream to the Constructor is difficult?  I don't think I understand the case you are concerned with.

Another nice thing about passing ReadableStream to Constructor is that its clear fixed bodies are immutable.  For example, it avoids this kind of confusion:

```
var r = new Request(url, {body: arrayBuffer});
r.body.write(...);   // does this append to the arraybuffer?
```

**Comment by tyoshino** [comment]
@wanderview `r.body` doesn't have `.write()`. Tentative plan is to reveal a writable byte stream via RequestInit. See https://github.com/slightlyoff/ServiceWorker/issues/413#issuecomment-52291880

**Comment by domenic** [comment]
@wanderview sure. Here's one scenario: I want to upload `chunk1` and `chunk2` and `chunk3` to `https://example.com`, and I want to know when each chunk has finished uploading (as well as when the whole upload process is finished, including any trailers or shutdown or similar). I want this so I can display accurate progress, ensure transaction-ish guarantees, and so on.

With a writable stream, this is easy:

```js
fetch('https://example.com', {
  body(ws) {
    ws.write(chunk1).then(() => updateProgress(1/3));
    ws.write(chunk2).then(() => updateProgress(2/3));
    ws.write(chunk3).then(() => updateProgress(3/3));
    
    ws.closed.then(showUploadComplete, showUploadFailed);
    
    // NB: we will probably have writers in the final writable stream API.
    // Ignoring them for now.
  }
});
```

I don't see how you would do this with the pass-in-a-readable stream version.

> Another nice thing about passing ReadableStream to Constructor is that its clear fixed bodies are immutable. For example, it avoids this kind of confusion:

I don't think `r.body` should be a writable stream, so that example doesn't make much sense to me. Only the creator of the `Request`/`Response` should get a writable stream.

**Comment by annevk** [comment]
Does
```js
var req = new Request("https://example.com", {
 body(ws) { ... }
})
fetch(req)
```
also work?

**Comment by wanderview** [comment]
@domenic isn't this equivalent to your example?

```
fetch('https://example.com', { body: new ReadableStream({ pull: function(controller) {
  var i = ~~this.chunkIndex;
  controller.enqueue(chunks[i]);
  updateProgress((i + 1) / 3);
  this.chunkIndex = i + 1;
  if (this.chunkIndex == 3) {
    controller.close();
  }
});
```

Slightly more code, but you get back-pressure.  The problem with just exposing WritableStream is it makes is more likely back pressure is forgotten.

So it seems no matter what we end up with the concept of a pipe here.  An object with ReadableStream and WritableStream interfaces with a buffer.

Either we can have Request instantiate the pipe and then expose the writer to the creator of the Request, or we can have the creator instantiate the pipe and pass the ReadableStream.

I guess I favor having the script instantiate the pipe and pass the ReadableStream because you can then avoid the whole pipe/buffer if you already have a ReadableStream.

A pipe mock up:

```
var pipe = new Pipe();

pipe.writer.write(chunk1).then(() => updateProgress(1/3));
pipe.writer.write(chunk2).then(() => updateProgress(2/3));
pipe.writer.write(chunk3).then(() => updateProgress(3/3));
pipe.writer.closed.then(showUploadComplete, showUploadFailed);

fetch('https://example.com', { body: pipe.reader });
```

**Comment by domenic** [comment]
@wanderview no. There's no way know when the chunk is finished uploading. Your updateProgress will fire too fast.

The pipe example has the same problem. The writes will complete as soon as the readable stream accepts the chunk. Whereas, we want to know when the HTTP stack has accepted the chunk. There's no way to extract that information, since the code that reads from `pipe.reader` doesn't have a way to communicate it back to `pipe.writer`.

**Comment by domenic** [comment]
@annevk yeah, it should.

**Comment by tyoshino** [comment]
In other word, WritableStream-UnderlyingSink pair communicates data and ack while ReadableStream-UnderlyingSource pair communicates only data. In #287, I was prototyping an interface named OperationStream which has WritableStream and modified ReadableStream which allows the consumer to acknowledge completion of consumption of a data read from the ReadableStream.

Especially, this will be problematic when we want to reuse a buffer passed to the sink. As current ReadableStream-UnderlyingSource pair cannot tell when the sink has done consuming all the bytes in the given ArrayBuffer, so the underlying source cannot return the buffer to the user for reuse. `WritableStream.prototype.write(data)` can notify the producer when `data` has been consumed by the sink.

**Comment by wanderview** [comment]
> @wanderview no. There's no way know when the chunk is finished uploading. Your updateProgress will fire too fast.

I don't see how you can accomplish this with a Request object that can be passed anywhere (fetch, cache, random script).  All of these consumers are just going to be reading the body as a stream.

Unless you are suggesting fetch should do something magical that is not represented in the stream API?

**Comment by tyoshino** [comment]
@wanderview, that's a good point. As I said in https://github.com/yutakahirano/fetch-with-streams/issues/30#issuecomment-90137980, the sink (cache, fetch, etc.) cannot tell the Request to propagate successful completion of writing/sending/etc. via `req.body`. It requires yes, magical intrusive operation on the Request.

**Comment by wanderview** [comment]
At a conceptual level, I'm not sure it makes sense to try to make a source stream express downstream progress of its consumer.  It feels like trying to force a square peg into a round hole to me.  I think something like this is a better fit:

```
  fetch(request, { progress: handler });
```

**Comment by domenic** [comment]
>  All of these consumers are just going to be reading the body as a stream.

That wasn't how I envisioned this.

> Unless you are suggesting fetch should do something magical that is not represented in the stream API?

No, not at all. Here, let me demonstrate one quickly-hacked-together implementation that should do the trick:

```js
if (typeof requestInit.body === "function") {
    let resolveIsStarted;
    const isStarted = new Promise(resolve => resolveIsStarted = resolve);

    let dest;

    const bodyWS = new WritableStream({
        start() { return isStarted; },
        write(chunk) {
            return isStarted.then(() => {
                if (dest === "cache") {
                    return write(cacheFD, chunk);
                }
                if (dest === "upload") {
                    return send(socketFD, chunk);
                }
            });
        }
    });

    requestInit.body(bodyWS);
}

// adding to a cache: set `dest = "cache"` and call `resolveIsStarted()`
// uploading: set `dest = "upload"` and call `resolveIsStarted()`
```

This kind of adapter is what I was getting at with my OP.

This doesn't handle letting script read from `request.body`, but you could hook up something pretty easily... something like

```js
const bodyRS = new ReadableStream({
    pull() {
        dest = "readablestream";
        resolveIsStarted();
    }
});
```

and then add another `if` inside `write`. (Hmm, this might not work with `pull` being called immediately after the readable stream finishes starting... maybe we should wait for a `getReader()` call?)

---

Re the last post: at a conceptual level, I don't see how it makes sense to fit the square peg of readable streams into the round hole of uploading/writing to the cache.

**Comment by wanderview** [comment]
> Re the last post: at a conceptual level, I don't see how it makes sense to fit the square peg of readable streams into the round hole of uploading/writing to the cache.

I would agree with this statement if we had `cache.write()` and `fetch.write()`, but we don't.  We have a buffered object that is passed to cache and fetch.

Having a hard coded list of "if consumer is this type then do this" conditionals in fetch Request sets alarm bells off for me.

Is there way for a js library-defined consumer to take part in this behavior?  It seems not.

**Comment by domenic** [comment]
Then I think we need to start talking about design changes to fetch/cache that will allow writing directly. If you don't think the request/response object is a good way to expose this, we need something better.

**Comment by tyoshino** [comment]
Domenic, maybe Ben called code like you just introduced as "magical". It's no longer using the publicly facing APIs of the Request. We're peeling the envelope of the Request to expose its internals and attaching consumer (cache, fetch, etc.).

**Comment by wanderview** [comment]
That seems reasonable.  Are there other reasons to want a consumer-specific WritableStream here besides hooking progress event?

If its purely for progress, it seems we might want something that supports getting progress for the upload of fixed bodies, like `new Request(url, { body: giantArrayBuffer })`, as well.

**Comment by tyoshino** [comment]
I investigated if we can delay fulfillment of the promise returned by .write() until the data is read from the ReadableStream to address this issue. When ReadableStreamReader.read() was specced to be synchronous, we could use this approach. The only limitation is that the consumer must consume or copy all the bytes from the ArrayBuffer returned by read() synchronously. This allows the consumer (cache, fetch, etc.) to tell when the data gets consumed indirectly to the data producing code that's using the writable side. Now, ReadableStreamReader.read() is asynchronous. So, we cannot use this approach. Extended version of ReadableStream like what I prototyped in https://github.com/whatwg/streams/pull/287 is necessary.

If we have this, completion of write can be propagated from the final consumer (cache, fetch) to the producer code through Request object.

**Comment by domenic** [comment]
> Are there other reasons to want a consumer-specific WritableStream here besides hooking progress event?

Definitely! It's all the reasons you want a WritableStream class to exist, instead of just saying your ecosystem is ReadableStreams + functions accepting them.

Most generally, you want to be able to give this writable stream to generic code that deals with writable streams.

> If its purely for progress, it seems we might want something that supports getting progress for the upload of fixed bodies, like new Request(url, { body: giantArrayBuffer }), as well.

I think progress, in whatever form, needs to be built on top of the underlying streaming primitives. Since we've just abandoned Request/Response as the streaming primitives, we need to figure out what lies underneath them before we can talk about progress...

**Comment by wanderview** [comment]
> Definitely! It's all the reasons you want a WritableStream class to exist, instead of just saying your ecosystem is ReadableStreams + functions accepting them.

Where is that written up?  Sorry if I missed it above.

> Most generally, you want to be able to give this writable stream to generic code that deals with writable streams.

Again, I don't see why a pipe construct can't resolve this issue with the current Request/Response.

> I think progress, in whatever form, needs to be built on top of the underlying streaming primitives. Since we've just abandoned Request/Response as the streaming primitives, we need to figure out what lies underneath them before we can talk about progress...

Well, I think @annevk should chime before declaring Request/Response dead as streaming primitives. :-)

As I mentioned on IRC, I expect browsers don't really report "bytes written to the wire" for progress today.  They probably just report based on internal buffer consumption.  I think this would be consistent with reading from the Request body stream.

**Comment by domenic** [comment]
> Where is that written up? Sorry if I missed it above.

Oh, I kind of assumed it was clear that we wanted a primitive that wraps underlying data sinks into a generic interface that libraries and authors could operate on, with backpressure etc.

So you think instead of

```js
const ws = fs.createWriteStream("/path/to/file.flac");
ws.write(generateHeaders());
rs.pipeTo(ws).then(() => ws.write(generateTrailers()));
ws.closed.then(showSuccessUI, showFailureUI);
```

we should do something like

```js
fs.writeToFile("/path/to/file.flac", new ReadableStream({
  start(c) {
    c.enqueue(generateHeaders);
    pumpRS(c).then(() => c.enqueue(generateTrailers()));
  }
})
.then(showSuccessUI, showFailureUI);
```

?

> Well, I think @annevk should chime before declaring Request/Response dead as streaming primitives. 

To clarify, by "streaming primitives" I mean "exposing directly the ability to do uploads or cache writes in a streaming way".

> They probably just report based on internal buffer consumption. I think this would be consistent with reading from the Request body stream.

Agreed they would be consistent. So I guess for XHR there's no way to know if an upload succeeded, besides just the server giving back 2xx? I guess that makes sense. Doesn't work for filesystem streams though...

**Comment by wanderview** [comment]
I am incorrect about XHR.  At least in gecko I'm told by our network gurus we report progress based on bytes sent to the kernel.  OS buffering means progress is really only meaningful for very large uploads, though.

**Comment by tyoshino** [comment]
See http://krijnhoetmer.nl/irc-logs/whatwg/20150406#l-775 for discussion happened in IRC.

**Comment by annevk** [comment]
> To clarify, by "streaming primitives" I mean "exposing directly the ability to do uploads or cache writes in a streaming way".

Is the problem that these need different primitives? Why is that? Can't we provide the underlying source the moment `Request` is passed to a cache or a network operation?

**Comment by domenic** [comment]
Progress report from IRC ([logs](http://krijnhoetmer.nl/irc-logs/whatwg/20150406#l-775)):

One potential plan would be to explain things like this:

- Passing in a ReadableStream to bodyInit will work.
- `Request` is minimally extended to allow a destination writable streams to be "set" in certain situations. [@wanderview's gist](https://gist.github.com/wanderview/ac6052184c62d2f165ee) gives a sketch.
- UAs have "hidden" magic writable streams, call them e.g. CacheWritableStream and HttpUploadStream. When you do `fetch(req)`, it does approximately `req.setWriter(new HttpUploadStream(req))`: that is, it creates a `HttpUploadStream` that is connected to the server as specified by the details in `req`, and sets `req.bodyWriter` to that stream. When you do `cache.add(req)`, similarly for `CacheWritableStream`.
- For now we don't expose the ability to directly create `HttpUploadStream`s or `CacheWritableStream`s. Maybe we could later if someone asks for them. But for now they're indirectly created via fetch/cache.
- This is highly extensible, since authors can create their own writable streams and call setWriter if they want to redirect a response somewhere else. For example... let's say that someone created a video element with a writable stream, `videoEl.writable`, that you can write bytes in to. You could then write a method `videoEl.appendRequest(req)` which does internally `this.setWriter(this.writable)`. That's kind of neat, and lets us feel like we're not hard-coding specific behaviors into Request.

**Comment by tyoshino** [comment]
This doesn't looks so good. The WritableStream exposed via `req.bodyWriter` is no longer a part of the `Request` object. `Request` is now like a washer through which a bolt of the WritableStream is exposed to the user code.

If we're going not to have a buffer in a Request/Response by default, I think the WritableStream revealer should be part of the final destination API. For example, fetch() taking FetchInit as a third parameter and we pass the revealer in FetchInit. It looks strange to me that the Writer is exposed as above.

----

I find it a little inconvenient that we cannot start piping until fetch() is invoked. If there's already started push source, we need to have a queue (e.g. wrapping the push source using ReadableStream library, or have a queue with readable side and writable side and push generated data to the queue, ...) separate from a Request object to buffer the data until we're ready to invoke `fetch()`. If Request has a WritableStream with a queue in it by default, we can just push data to it until we're ready to call `fetch()`.

As I have been proposing, we can choose to optionally expose an extended ReadableStreamReader on Request/Response whose read() returns a promise of a tuple of { value, resolve, reject }`, so that commit to file, write to kernel, etc. can be propagated to the user via WritableStream side of the Request/Response. We can have this as option so that the basic read() method that just returns a promise of a value is kept as-is.

**Comment by yutakahirano** [comment]
>> Re the last post: at a conceptual level, I don't see how it makes sense to fit the square peg of readable streams into the round hole of uploading/writing to the cache.

> I would agree with this statement if we had cache.write() and fetch.write(), but we don't. We have a buffered object that is passed to cache and fetch.

Yeah, some changes are needed for the fetch algorithm to support upload streaming. It means we will have fetch.write() in the future.

**Comment by annevk** [comment]
When you create a `Request` from an existing `Request` (which is what `fetch()` does), it simply transfers the *writable stream* from one object to the next.

I still think that optimizing the piping as proposed in https://github.com/yutakahirano/fetch-with-streams/issues/30#issuecomment-90151618 is sensible.

However, if we don't want that we should instead use the revealing pattern. You pass a callback to `Request`. Then once some party starts reading that stream (`.body.getReader()`, `fetch()`, `cache.put()`) you reveal the writable stream after having coupled it to the underlying sink (perhaps after a microtask).

New API surface should really be a last resort and I think I'd rather reconsider the entire `fetch()` API at that point as it clearly wasn't good enough. Looping in @jakearchibald as he might have thoughts on this too.

**Comment by yutakahirano** [comment]
Regarding ack, I don't think fetch must ensure transaction-ish guarantees. One reason is ServiceWorker - I think we cannot guarantee anything. More generally, fetch can run on a virtualized environment: a page controlled by a ServiceWorker, a virtualized OS, a Chrome renderer process, etc. To which layer we should send data depends on the environment.
WritableStream looks fine because it enables the BYOB-style, though.

Text, ArrayBuffer, etc. can be read from Request and I would like to handle the writable stream on Request similarly. I don't think it's good to bind the stream to a specific consumer at the creation time. As we allow users to read Response.body partially and put the remaining to cache, we should allow users to read Request.body partially and put the remaining to fetch (in ServiceWorker, for example).

**Comment by annevk** [comment]
@yutakahirano so that argues for the solution from https://github.com/yutakahirano/fetch-with-streams/issues/30#issuecomment-90151618 which @wanderview does not like but might be okay with, right?

**Comment by tyoshino** [comment]
I also recall that we initially had `bool send()` on WebSocket but eventually made it `void` as it's underspecified what the bool value means. Data is accepted by kernel, sent IPC, buffer has been consumed, ... etc. etc.

I agree with Yutaka that there's no meaningful signal to return via the promise returned by `.write()` in the `fetch()`'s case.

(One of some sinks I can come up with for which the promise returned by `write()` may be useful is, a sequence of records to be committed to DB. Streams API works well for representing consumption of a sequence of records. We clearly define that fulfillment of the promise means that the record has been committed to the DB. But fetch() is not the case.)

**Comment by yutakahirano** [comment]
> @yutakahirano so that argues for the solution from #30 (comment) which @wanderview does not like but might be okay with, right?

Sorry for not being clear.

I'm for https://github.com/yutakahirano/fetch-with-streams/issues/30#issuecomment-90151618. But I don't think https://github.com/yutakahirano/fetch-with-streams/issues/30#issuecomment-90125685 is needed for the fetch API. I'm against https://github.com/yutakahirano/fetch-with-streams/issues/30#issuecomment-90198610.

**Comment by tyoshino** [comment]
To be clear, I'm:
- fine with making Request constructor accept a ReadableStream
- fine with making something magical to give control over the behavior of the WritableStream of a Request to `fetch()` like https://github.com/yutakahirano/fetch-with-streams/issues/30#issuecomment-90151618 though I'm still investigating making it more simple by using ReadableStream with ack functionality
- against making Request non-writable until the Request is passed to `fetch()`, etc.

**Comment by wanderview** [comment]
I think this actually works pretty well if we treat Request/Response having their own `pipeTo()` functions.  These are shorthands for .body.getReader().pipeTo() in most cases, but if a ws-revealer function is provided for the body then it gives direct access to the consumer pipe.

Mock up:

```
// create a fixed ReadableStream for body
var req1 = new Request(url, { body: arrayBuffer });

// access fixed ReadableStream directly
req1.body.getReader()

// or lock ReadableStream and pipe to consumer WritableStream
// shorthand for req1.body.getReader().pipeTo(consumerWritableStream)
req1.pipeTo(consumerWritableStream);

//----------------------------------------------------

// set body to given ReadableStream
var req2 = new Request(url, { body: ReadableStream });

// access passed ReadableStream directly
req2.body.getReader()

// or lock ReadableStream and pipe to consumer WritableStream
// shorthand for req2.body.getReader().pipeTo(consumerWritableStream)
req2.pipeTo(consumerWritableStream);

//----------------------------------------------------

function writeBodyFunc(ws) { ... }

// body stream is not defined yet... depends on how accessed below
var req3 = new Request(url, { body: writeBodyFunc });

// on first getReader() call, construct a pipe and pass to writeBodyFunc to populate,
// return pipe reader
req2.body.getReader()

// or call writeBodyFunc() with consumer stream directly, treat getReader() as locked
req2.pipeTo(consumerWritableStream);

```

Things like Fetch and Cache would be spec'd to call Request.pipeTo() and Response.pipeTo() with implementation specific WritableStream objects.  JS consumers could also take part by calling .pipeTo() as well.

Thoughts?

**Comment by tyoshino** [comment]
What I didn't like is that `writeBodyFunc` gets called only after the `Request.pipeTo(finalConsumer)` or the `req2.body.getReader()` happens. I think Request/Response should become able to be populated through a WritableStream immediately after construction, and start accepting `write()`s for convenience. Once `Request.pipeTo(finalConsumer)`, etc. happens, you could do some optimization internally to start skipping the buffer inside Request.

**Comment by domenic** [comment]
Haven't had time yet today to fully digest all of the things in this thread, but just chiming in a bit with first impressions for what they're worth...

I don't quite like the pipeTo idea. I'd rather keep stream things inside the actual stream. setBodyWriter sounds better.

I think I'm pretty OK with delaying the call to `writeBodyFunc`... it doesn't seem too bad to me that you can't write to an "in memory" writable stream, and instead you only get a writable stream when there's an actual corresponding underlying sink ready.

I'm worried about statements like

> I agree with Yutaka that there's no meaningful signal to return via the promise returned by .write() in the fetch()'s case.

I think "accepted by the kernel" (or virtualized kernel, or whatever) is reasonably meaningful... But I don't have as much I/O experience as you guys, probably. If you tell me this really isn't something worth surfacing, I'll believe you. The fact that XHR does use that signal seems like a counterpoint. Then again, maybe we shouldn't be distinguishing between the kernel buffer and the user-space buffer.

If we really think this is not meaningful, we should reconsider this aspect of writable streams more generally. Currently they are a bit complicated because I thought it was paramount to give both backpressure signals and has-this-write-succeeded signals. If you tell me the latter signal is not important, we can simplify. I am not sure that cases like

> One of some sinks I can come up with for which the promise returned by write() may be useful is, a sequence of records to be committed to DB.

are compelling enough...

@tyoshino's ReadableStream-with-ack is also something I need to think about more. Adding more members to the `{ value, done }` tuple is a reasonably painless way of doing this that helps my ergonomics concerns. But I just haven't sketched it out in enough detail to tell whether it gives us something worthwhile (besides the potential for symmetry---symmetry with a model that might be flawed anyway, as we note above).

Hope this helps a bit.

**Comment by tyoshino** [comment]
> I think "accepted by the kernel" ...

Let's think of what we can do in response to the signal. It doesn't mean the data has arrived at the destination server. It just means the data reached at some point in the long path to the final application server code. If the user is keeping the data in case it gets lost on the network, what the user want to receive is a message saying "ok, the bytes are successfully processed" from the peer server's application layer code. It's not "I'm now travelling in kernel".

The "Hey user, I'm travelling in kernel now" signal is meaningful only for the backpressure perspective.

**Comment by tyoshino** [comment]
> If we really think this is not meaningful, ...

I do think it's meaningful for some non-fetch cases such as one I introduced in the parentheses in https://github.com/yutakahirano/fetch-with-streams/issues/30#issuecomment-90502696.

**Comment by domenic** [comment]
I guess I am most curious about whether it is ever meaningful for I/O cases. For example for files you have the same problem, right? Assuming we don't call flush(), knowing the kernel buffer has accepted your write is not useful. If we say it's only useful for non-I/O streams, I think that should give us pause. 

(The database record case is interesting since it's somewhere midway between a pure "I/O" stream and a higher-level object stream.)

How often will cases where meaningful acknowledgement is possible pop up on the web platform? If the answer is "very rarely," maybe those should be handled by domain-specific streams (e.g. DBStream inherits WritableStream and includes a `.writeWithCompletion() -> { backpressurePromise, completionPromise }` method for use by its specific consumers, or something).

**Comment by tyoshino** [comment]
> I think I'm pretty OK with delaying the call to writeBodyFunc... it doesn't seem too bad to me that you can't write to an "in memory" writable stream, and instead you only get a writable stream when there's an actual corresponding underlying sink ready.

If the WritableStream gets that semantics, it seems the revealing function should be a part of the Fetch, Cache, etc. as I said in: https://github.com/yutakahirano/fetch-with-streams/issues/30#issuecomment-90352974. I don't see much value to have it appear on Request.

**Comment by domenic** [comment]
> If the WritableStream gets that semantics, it seems the revealing function should be a part of the Fetch, Cache, etc.

That makes sense to me too. It's easy to think of Request as "an options bag for fetch" but as we've shown in this thread that's not really the case.

**Comment by wanderview** [comment]
If you want to write to a Request before its been passed to a consumer, then you are asking to buffer your data.  In this case, just explicitly write to a pipe (aka buffer) and pass the reader side as the body.

**Comment by tyoshino** [comment]
@wanderview Given that we're likely to conclude that we make the Request constructor accept ReadableStream, I'm fine with that. Then, we won't add WritableStream revealing function overload to the Request constructor, but instead we add an option item (e.g. FetchInit) to pass WritableStream revealing function directly to the final sink (e.g. fetch()).

**Comment by wanderview** [comment]
> @wanderview Given that we're likely to conclude that we make the Request constructor accept ReadableStream, I'm fine with that. Then, we won't add WritableStream revealing function overload to the Request constructor, but instead we add an option item (e.g. FetchInit) to pass WritableStream revealing function directly to the final sink (e.g. fetch()).

Ok, but that seems to have problem that now you have to deal with what to do when there is a body on the Request and also a fetch ws-revealing function.  I guess that's solvable, but it seems a bit odd to me.

In all of the ws-revealing function proposals, though, how do we plan to handle tee() operations in the fetch algorithm when it has to do redirects?

**Comment by wanderview** [comment]
I guess to be more clear, I think it would be preferable to keep the Request as a complete description of a network request instead of trying to move part of it to one of the consumers of that request.

**Comment by tyoshino** [comment]
I tried to get convinced with the idea.

I was thinking that the Request and Response are representing data, not operation. This point of view made sense when I was thinking of fetch canceller (https://github.com/whatwg/streams/issues/297#issuecomment-81410226). But it seems I shouldn't be dragged with it when thinking of writable stream revealer. The writable stream revealer is generic to all of the body consuming operations.

The following is the (current) exhaustive list of operations that consumes body.
- (a) Doesn't set body passed: `clone()`, `arrayBuffer()`, `blob()`, `formData()`, `json()`, `text()`
- (b) Sets body passed: `new Request()` (includes `fetch()`), `cache.put()`, `e.respondWith()`

They are grouped based on whether it affects body passed flag or not. See https://github.com/tyoshino/streams_integration/blob/master/FetchBodyPrecondition.md#a-lock--body-passed-flag. 

For Request/Response constructed with a writable stream revealer:
- the operations in the group (a):
    - sets up a queue and consume (really consume, or just tees) from its readable side if it's not yet set up and reveal the writable side through the revealer, or
    - consumes from the queue which has been set up by the first call of any of the operations in the group (a)
- the operations in the group (b):
    - exposes their own WritableStream through the revealer, or
    - pipes from the queue (which has been set up by the previous call of any of the operations in the group (a)) to their own WritableStream

The requirement that we neuter the body with `body passed` flag when any of the group (b) operations is called ensures that writable stream revealing happens only once.

The first body consuming operation interprets the `RequestInit.body` and `BodyInit`, and the later body consuming operations see a ReadableStream.

Above impresses me that RequestInit.body and BodyInit are not so bad places to hold a writable stream revealer.

**Comment by tyoshino** [comment]
(Cont'd) `new Request()` operation doesn't have its own WritableStream. As Anne said, it has been supposed to detach body from the old Request. If the old Request was constructed with a writable stream revealer, maybe we just transfer the ownership of the revealer to the new one, and the new one behaves as if it was constructed with a writable stream revealer. I think it just works.

**Comment by tyoshino** [comment]
One thing I found unsure is in chat with Yutaka is:

```
let r0 = new Request(url, { body: revealer0 });
let r1 = new Request(r0, { body: revealer1 });
let p = fetch(r1);
```

For non revealer body, current fetch spec gives higher priority to _init_'s body. So, for this case, similarly, I think `revealer1` should win. So, `revealer0` would either:
- never gets called
- get a writable stream to nil (like /dev/null)
- get an errored writable stream

**Comment by tyoshino** [comment]
After thinking as above, I'm a little less uncomfortable with Ben's idea. But the point that a Request reveals a WritableStream not on construction but only when it starts being consumed still looks not easy to understand and looks against this principle https://github.com/whatwg/streams/blob/master/Requirements.md#you-must-be-able-to-efficiently-adapt-existing-push-based-data-sources-into-a-uniform-streaming-interface.

**Comment by yutakahirano** [comment]
> I guess to be more clear, I think it would be preferable to keep the Request as a complete description of a network request instead of trying to move part of it to one of the consumers of that request.

I agree with it, but your proposal looks binding the writable stream to one consumer and I don't like that. I still don't understand why [Domenic's code] (https://github.com/yutakahirano/fetch-with-streams/issues/30#issuecomment-90151618) is bad. The model is easy to understand, and we can optimize the implementation for typical cases. Do you think your proposal is easier to understand? Or is it more efficient?

**Comment by wanderview** [comment]
> I agree with it, but your proposal looks binding the writable stream to one consumer and I don't like that. I still don't understand why Domenic's code is bad. The model is easy to understand, and we can optimize the implementation for typical cases. Do you think your proposal is easier to understand? Or is it more efficient?

If you change it so that js libraries can provide the consumer WritableStream then I'm ok with it.  Which is basically what my proposal did.  Just change Domenic's code to wrap the consumer set WritableStream instead of using a magic enumeration switch:

```
if (typeof requestInit.body === "function") {
    let resolveIsStarted;
    const isStarted = new Promise(resolve => resolveIsStarted = resolve);

    let dest;

    const bodyWS = new WritableStream({
        start() { return isStarted; },
        write(chunk) {
            return isStarted.then(() => {
                return dest.write(chunk);
            });
        }
    });

    // expose a function to let cache, fetch, or js set the consumer WritableStream
    this.pipeTo = function(ws) {
      dest = ws;
      resolveIsStarted();
    };

    requestInit.body(bodyWS);
}
```

Then cache, fetch, or any js consumer can call request.pipeTo() to set the specific consumer stream.  The internal wrapper stream is unblocked at that point.

**Comment by domenic** [comment]
Let me see if I understand the current proposal correctly:

- `request.setWriter(writableStream)` exists publicly, and is also called internally by `fetch` and `cache.add`.
- `new Request({ body(ws) { /* do stuff */ } })` will `/* do stuff */` either:
   - after `setWriter` is called, or
   - immediately, using an adapter writable stream to buffer while waiting for a writer to be set (as in [Ben's example](https://github.com/yutakahirano/fetch-with-streams/issues/30#issuecomment-90985633)).
    - which is it? I am OK with either, maybe leaning toward the lazy initialization that doesn't require a wrapper?
- `new Request({ body: readableStream })` works; it is sugar for `new Request({ body(ws) { readableStream.pipeTo(ws); } })`.

Things I'm not clear on:
- Assuming `request.body` will be a Readable(Byte)Stream, do we say that accessing the `body` property does something like `const pipe = new TransformStream(); this.setWriter(pipe.writable); return pipe.readable;`? That's a reasonably big side effect for a getter to have, hmm. Or can we delay that until they call `getReader()`? That sounds better, although I think it will require some spec work for me on Streams.

**Comment by wanderview** [comment]
I think we're still on different pages.

You seem to like `setWriter()` as a name better than `pipeTo()`, but it needs to be more than just setting a value.  This call must also say "lock the body and drain it to this WritableStream".  To me that is what pipeTo() does, but we can call it drainToWriter() or something if you want.

I would prefer not using a wrapper for the ws-revealer function since it might interfere with possible underlying source-to-sink optimizations.  But I don't think that is a huge concern here since there are other options available for passing the body.

I think `new Request({ body: readableStream})` should just set that stream as the body.  The `setWriter()` or `drainToWriter()` call would then do `body.pipeTo(ws)` instead of calling any ws-revealer function.

This is different from what you wrote above, because there is no object creation or magic needed to support `request.body.getReader().read()` when passing a body:readableStream.  The body is just the stream provided; period.

The magic that is needed is when a ws-revealer function is used, but someone calls `body.getReader()` instead of `setWriter()` or `drainToWriter()`.  In this case, we must create a pipe, pass the writer side to the ws-revealer, and then make the body getter return the pipe reader side.

**Comment by tyoshino** [comment]
> I would prefer not using a wrapper for the ws-revealer function since it might interfere with possible underlying source-to-sink optimizations. But I don't think that is a huge concern here since there are other options available for passing the body.

I wrote some possible solution for this issue at #325.

**Comment by domenic** [comment]
@wanderview OK, I see. I was unsure what the integration was between supplied bodies and the writable stream. Let me try to re-summarize.

- `Request` constructor accepts a `body` that can be a "chunk" (array buffer etc.) or a readable stream or undefined
  - If a readable stream, it's left as-is for `req.body`
  - If a chunk, a new readable stream is created as `req.body` and that chunk is enqueued
  - If undefined, `req.body` is undefined.
- `Request` constructor accepts a `bodyWriter` function.
  - This can be supplied along with a `body` option; there is no conflict
- `req.drainToDestination(writableStream)` exists publicly, and is also called internally by `fetch` and `cache.add`.
  - Upon calling, it does a few things.
  - First, if `this.body` is not undefined, it does the equivalent of `this.body.pipeTo(writableStream, { preventClose: /* true iff bodyWriter was suplied */ })`. (This locks `this.body` and it won't be unlocked until it is drained.)
  - After that finishes, if `bodyWriter` was supplied, it calls `bodyWriter(writableStream)`.

BTW the reason I am unsure about request.pipeTo is I don't like the confusion between it and request.body.pipeTo. Even if they are conceptually similar, as we can see above they do different things.

**Comment by wanderview** [comment]
I agree with most of that, but I don't like that Request.body.getReader() fails if you only supply a bodyWriter function to new Request().

It seems the consumer of the Request should not have to understand how the request for was created.  It should be be able to read them regardless.

**Comment by domenic** [comment]
Hmm, I was trying to avoid the magical pipe, but I see your point it's a strong one. And actually upon revising (see above edits with \<ins\>/\<del\>) it seems there's still no magical pipe-creation necessary. WDYT?

**Comment by wanderview** [comment]
I don't think that is adequate unless we want to say "you *must* use drainToDestination() and not rely on .body.getReader() directly".  Because otherwise you miss the body data represented by the bodyWriter function.

Btw, I prefer to just pass the bodyWriter function as the body: property, but I won't insist if you prefer a separate bodyWriter: property.

**Comment by tyoshino** [comment]
> If a chunk, a new readable stream is created as req.body and that chunk is enqueued

Nitpick: The chunk is converted into ArrayBufferView and enqueued.

**Comment by tyoshino** [comment]
> If undefined, req.body is undefined.

This should just create a closed ReadableStream.

**Comment by tyoshino** [comment]
Clarification: I mean, if `RequestInit.bodyWriter` is undefined && `RequestInit.body` is undefined, `req.body` should not be a closed ReadableStream. If `bodyWriter` is supplied via `.body`, it's ok.

**Comment by tyoshino** [comment]
How about making the reveal-writable-stream-of-fetch functionality optional? I.e. `RequestInit` has an entry named `noqueue`. If its set to **true**, revealWritableStream function passed to `RequestInit.body` returns the `WritableStream` of `fetch()`, `cache.put()`, and therefore the function is invoked only when `fetch()`, `cache.put()` is invoked. This is for performance sensitive application. If `noqueue` is set to **false** (default), the revealWritableStream function is called synchronous to `Request` construction and a queue (Pipe in Ben's terminology) is set and its writable side is revealed.

Counter argument to this approach is "oh, why? you can just create a Pipe by yourself and pass it to `RequestInit.body`". But it depends on whether one thinks the `noqueue` == **false** behavior (considered to be convenient and straight by me and @yutakahirano) is more common or the `noqueue` == **true** behavior (considered to be important for performance by @wanderview) is more common.

**Comment by yutakahirano** [comment]
@wanderview thanks, I think my understanding is better than before, so let me describe it. Correct me if I'm wrong.

Let `rs` and `ws` be the streams associated to `req`, a Request. `rs` is `req.body` and `ws` is the stream to be revealed. `rs` and `ws` have a source `src` and a sink `sink`, respectively.

--

Your idea is that we should create `ws` when we have `sink`, e.g. calling `cache.put`, calling `fetch`, etc, and the reason is that we don't have a means to change the source or sink after stream construction. One problem is that we have `rs` (a.k.a. `req.body`) so it is possible to read data partially from `req.body` and then call `fetch(req)`. This can be solved by making the consumer read from `rs` in this case. `ws` doesn't know the actual consumer any more (i.e. pipeTo is not optimized in this case), but it's bearable.

--

I thought we could use one sink that can handle multiple consumers, but I now think it is not very attractive because a sink should correspond to one consumer. That said, I still think we can optimize `rs2.pipeTo(ws)` if `src2`, the source of another readable stream `rs2`, is created by the platform. Because the platform created `sink`, it can directly drain data from `src2` and push it to `sink`. `sink` can set `ws.state` to `waiting` while the consumer is not determined so that one who wants to write into `ws` can notice that `ws` is not wanting data now.

--

The solutions look relying on an assumption that once the final consumer is set, it will never change. That may be correct for Request / Response, but I imagine that may be wrong in other use cases and  providing a means to change the source and sink is more general and preferable. https://github.com/whatwg/streams/issues/325 is a similar attempt, right?

**Comment by domenic** [comment]
@wanderview

> I don't think that is adequate unless we want to say "you must use drainToDestination() and not rely on .body.getReader() directly". Because otherwise you miss the body data represented by the bodyWriter function.

So, I guess while writing up the above, I realized that it's not accurate to say that the bodyWriter represents the body. It represents the action to take when this request is made somewhere that actually has a writable stream. The `Request` object doesn't actually have a body. Remember that a `Request` is "a complete description of a network request". In this case the description is "when a network request gets committed somewhere, this function will be responsible for writing out the body."

As such maybe the correct approach is just to expose `req.bodyWriter`. Then people who want to do what fetch/cache do can literally write an algorithm:

```js
function writeToMyDest(req) {
  return req.body.pipeTo(myDest.writable, { preventClose: req.bodyWriter })
    .then(() => req.bodyWriter(myDest.writable));
  
  // Here I assume the default value for bodyWriter is a no-op function.
}
```

> Btw, I prefer to just pass the bodyWriter function as the body: property, but I won't insist if you prefer a separate bodyWriter: property.

I hope the above explanation shows why I kept them separate. There's actually two things in a Request: a pre-computed body prefix (`body`), and a will-write-more-body function (`bodyWriter`).

---

Stepping back for a minute, and reviewing the latest comments from @tyoshino and @yutakahirano, I feel like we may have gotten a bit lost in the weeds on how to make every edge case work. Here's a higher level view of what I think we really want:

- `body: chunk`.
  - Will write `chunk` to destination writable stream, whenever Request is sent somewhere interesting.
  - Is readable out of `req.body` stream.
- `body: readableStream` for convenience.
  - Will pipe `readableStream` to destination writable stream, whenever Request is sent somewhere interesting.
  - Is readable as `req.body`.
- Some way of allowing direct access to the destination writable stream for fetches (important) and caches (less important).

Where we are getting stuck is, while trying to solve this last one, we are not being clear on what the requirements are with relation to `req.body`.

My proposal is that we keep these fairly logically separate. I'm not insistent on that though, and pretty much everyone else in the thread seems to want something more seamless, so that we abstract away the notion of "destination writable stream" into something more like "request body writable stream" that you can start writing to ASAP. That seems overcomplicated to me: I think most upload cases won't in general try to read from `req.body`, and will be happy to ignore `body` and just use a lazily-called `bodyWriter`. But I am largely speculating.

**Comment by wanderview** [comment]
If we are already planning to add some progress event system for the fixed body case, I think we should just give up on exposing a consumer-specific stream.

This gives you two layers of progress:

1) Bytes written to kernel via the progress event system.
2) Bytes read from body buffer via the streams API.

I think all the complexity here is from trying to force the stream progress system back to a time-disconnected object.  Its just not easily compatible.

That's what I would do if it was up to me, anyway. :-)

**Comment by domenic** [comment]
> I think all the complexity here is from trying to force the stream progress system back to a time-disconnected object. Its just not easily compatible.

Yeah, that's a concise statement of the problem.

I still think we really should have a way of exposing the fetch or cache writable stream some way. If we don't want to tie that to the time-disconnected Request object, that's OK. But I think it's important for streaming upload and such to work that way.

**Comment by annevk** [comment]
I think I'm okay with @domenic's approach. Offering a distinct API for the writable stream and keeping that distinct from the existing "chunk" approach.

Ideally I would have preferred a much higher-level streams primitive as e.g. evidenced by the Encoding Standard, but I've long since accepted that it wouldn't perform. So we should probably accept that readable and writable streams require slightly different API approaches.

**Comment by wanderview** [comment]
Personally I don't see why writing to a pipe is not adequate.  You can even use a pipe with a zero length buffer to get exactly the back pressure from the consumer.

Also, as an implementer I think I might oppose resolving .write() on "bytes written to kernel".  Its non trivial complexity in multi-process architectures and unnecessary if we are going to have a "progress event listener" style interface anyway.

**Comment by domenic** [comment]
> Personally I don't see why writing to a pipe is not adequate. You can even use a pipe with a zero length buffer to get exactly the back pressure from the consumer.

I guess this is true. Maybe it is the best marriage of a time-independent Request abstraction with the writable stream concept. I dunno.

> Also, as an implementer I think I might oppose resolving .write() on "bytes written to kernel".

I guess this is part of the larger discussion in https://github.com/whatwg/streams/issues/316? But then again, we could decide that the signal is generally useful but not worth implementing for HTTP streams. Everyone seems convinced it is not useful, so I won't insist on it.

It does seem rather horrible though that you could write code that is attempting to upload data somewhere, but you actually typoed the URL and you get a DNS error, but each call to `write()` just ends up fulfilling no matter what because they sit in the pipe.

**Comment by wanderview** [comment]
Do errors not propagate via pipeTo()?  I would expect a failed fetch to error it's writable stream which would propagate back to the Request pipe through the pipeTo() call.

Or if the fetch fails before pipeTo() it should be spec'd to cancel/close the Request body stream.

**Comment by domenic** [comment]
We're specifically saying there's a writable stream on one side of the pipe. That pipe is disconnected from the world during the period under contention, i.e. the "timeless" period before we send the request to cache or network.

What does the promise returned by write() do? For an ordinary stand alone pipe it would immediately fulfill... Are you thinking something different?

**Comment by domenic** [comment]
Oh I see, you were probably thinking of your proposal inhttps://github.com/yutakahirano/fetch-with-streams/issues/30#issuecomment-90985633 ? That would indeed take care of my concern.

**Comment by tyoshino** [comment]
In https://github.com/yutakahirano/fetch-with-streams/issues/30#issuecomment-92533398, it looks to me that Ben is talking about not write()-wise error but stream-wise error i.e. the error signal represented by `.closed`. Right? Ben.

Unless we want to return some error signal for each `.write()` (https://github.com/whatwg/streams/issues/316#issuecomment-91218528), it's fine to fulfill `revealedBodyWriter.write()` (here, `revealedBodyWriter` is not one of `fetch()`, but a WritableStream connected to the buffer in `Request`) immediately when it's queued in the buffer ("a pipe" in Ben's terminology) in the `Request`. If any error happens, `.closed` tells us that though it can be after fulfilling some of `.write()`s.

In Domenic's example in https://github.com/yutakahirano/fetch-with-streams/issues/30#issuecomment-92530561,

> that is attempting to upload data somewhere, but you actually typoed the URL and you get a DNS error

This is also a kind of "some error signal for each `.write()`" as well as "written-to-kernel" signal, I think.

Let's get on the same page. Domenic, do you think that `revealedBodyWriter.write()` should fulfill only after some part of the fetch algorithm steps go well? Do you have anything concrete in your mind, or just want to get prepared in case we want to propagate some error signal for each `.write()`?

For Fetch API's case, IMO it's ok to fulfill `.write()` immediately (i.e. before kernel, before DNS, before anything turns out to be successful).

**Comment by wanderview** [comment]
> Oh I see, you were probably thinking of your proposal in #30 (comment) ? That would indeed take care of my concern.

Right.  A pipe with zero buffer that provided back pressure until someone reads from it.  (I'll note this is much easier to do with async `.read()`.  With sync `.read()` we would need to do have a buffer of one chunk.)

**Comment by tyoshino** [comment]
Ah. Blocks fulfillment using isStarted. I see.

**Comment by domenic** [comment]
@wanderview 

>  A pipe with zero buffer that provided back pressure until someone reads from it.

Heh, I get confused by your usage of pipe... that comment shows a writable stream wrapping a writable stream, whereas I think of a pipe as a { readable, writable } pair ("identity transform stream"). Makes more sense now.

@tyoshino

>  Domenic, do you think that revealedBodyWriter.write() should fulfill only after some part of the fetch algorithm steps go well? Do you have anything concrete in your mind, or just want to get prepared in case we want to propagate some error signal for each .write()?

In light of https://github.com/whatwg/streams/issues/316#issuecomment-92541923 I have a very concrete answer for the minimum that I want: `write(ab)` should fulfill after it is safe to reuse `ab`. (Probably we would immediately detach `ab` upon calling `write(ab)`, and fulfill with an `ab2` that uses the same backing memory.) This might correspond to "written to the kernel," or it might correspond to "copied over IPC channels". Implementation dependent is probably OK since what we really care about is the memory here.

I also would like to catch as many errors as is feasible for each write. But I am not sure what capabilities we have in that regard. I think @wanderview's https://github.com/yutakahirano/fetch-with-streams/issues/30#issuecomment-90985633 (wrapper writable stream, I'll call it) would take care of this well. If anything causes the inner writable stream to error immediately (like a DNS failure), this will cause write() to reject based on isStarted rejecting. If anything causes the inner writable stream to error for an individual write, that will cause write() to reject based on the inner write() rejecting.

So I guess the question is, what does "If anything causes the inner writable stream to error for an individual write" mean. I.e., what is @wanderview comfortable communicating back to the main thread for an individual write. My initial hope was that we could get any of the errors send(2) would give: ECONNRESET in particular seems important to communicate back.  Does that seems possible?

**Comment by wanderview** [comment]
> Heh, I get confused by your usage of pipe... that comment shows a writable stream wrapping a writable stream, whereas I think of a pipe as a { readable, writable } pair ("identity transform stream"). Makes more sense now.

I do mean a Transform.  The write() does not resolve until someone calls read() to take the buffer.

**Comment by domenic** [comment]
In https://github.com/yutakahirano/fetch-with-streams/issues/30#issuecomment-90985633 there are no readable streams in sight. I am confused.

**Comment by wanderview** [comment]
I thought you were talking about the blocking concept in relation to that comment.  I thought we had moved passed the ws-revealer concept on the time disconnected object.

I am saying code can do something like this if they want a writable stream for the body:

```
var pipe = new Pipe(0 /* buffer size */);
startWritingMyStuff(pipe.writer);

var req = new Request(url, { body: pipe.reader });
fetch(req);
```

Here `pipe.writer.writer()` will not resolve until fetch calls `pipe.reader.read()` or `pipe.reader.pipeTo()`.  Conversely, if `pipe.reader.read()` is called first, it will not resolve until `pipe.writer.write()` is called.  With a zero length buffer the pipe flows pressure in both directions.

**Comment by tyoshino** [comment]
> should fulfill after it is safe to reuse ab

Yeah. For that purpose, I agree we should have some mechanism to propagate signal for each write(). Actually I've discussed that in the context of #323.

**Comment by domenic** [comment]
> I am saying code can do something like this if they want a writable stream for the body:

Oh. Hmm, I don't know how to implement that in terms of the current Readable and Writable stream APIs. The current model is that write()ing to a transform stream tells you only whether or not the transform stream accepted your chunk (e.g. it was a valid data format).

I guess we would need to make `controller.enqueue(chunk)` return a promise that is only fulfilled once that chunk has been `read()`, so that return value can be sent back to the `underlyingSource.write()` implementation. That is fairly sensible...

**Comment by tyoshino** [comment]
I guess this requires thin-stream as I prototyped before.

**Comment by domenic** [comment]
Right ... that's a big under the hood change, but shouldn't change public API or test expectations, hopefully.

**Comment by domenic** [comment]
Anyway, implementation aside, I am still confused what our general precedent is here. Are we saying that for e.g. a future FS API you should use it like 

```js
var pipe = new Pipe(0 /* buffer size */);
startWritingMyStuff(pipe.writer);

webFS.writeToFile(filename, pipe.reader);
```

?

**Comment by tyoshino** [comment]
I agree connecting underlyingSink.write() and dest.write() like the special pipeTo() in https://github.com/yutakahirano/fetch-with-streams/issues/30#issuecomment-90985633 works. But just `.read()` as-is doesn't work. At the point when `.read()` is called, we just started some async stuff (DNS resolution, write to kernel, IPC, etc.). If the `.read()` has ack functionality (https://github.com/whatwg/streams/issues/324), it works.

**Comment by tyoshino** [comment]
So, my comment https://github.com/yutakahirano/fetch-with-streams/issues/30#issuecomment-92967708 applies for how to implement the special `pipeTo()`.

**Comment by domenic** [comment]
I see. So @wanderview this is problematic. Your pipe design relies on adding more complexity for consumers of readable streams (https://github.com/whatwg/streams/issues/324). Otherwise given e.g.

```js
var pipe = new Pipe(0 /* buffer size */);
pipe.writer.write.then(onWritten);

var req = new Request(url, { body: pipe.reader });

// let's assume fetch(req), or the pipeTo call, or a manual consumer, does:
req.body.getReader().read().then(({ value, done }) => {
  doSomethingAsyncWith(value);
});
```

the code in `doSomethingAsyncWith(value)` expects the array buffer to be usable for some amount of time. We can't immediately detach `value` and call `onWritten` with a new ArrayBuffer with that same backing memory; we have to wait for doSomethingAsyncWith(value) to finish.

@tyoshino's proposed solution for this is something like

```js
var pipe = new Pipe(0 /* buffer size */);
pipe.writer.write.then(onWritten);

var req = new Request(url, { body: pipe.reader });

// let's assume fetch(req), or the pipe, or a manual consumer, does:
req.body.getReader().read().then(({ value, done, ack, ackError }) => {
  doSomethingAsyncWith(value).then(ack, ackError);
});
```

which would allow `onWritten` to be called once `doSomethingAsyncWith` finishes.

This might be necessary anyway to get usable transform streams, so I am glad we are exploring it now. But it is a pretty big departure from our current thinking on transform streams, which is that writing just immediately succeeds since it has been placed in the queue. (Transform streams have not received much love yet though, so I am not attached to the current design... just a little cautious.)

**Comment by wanderview** [comment]
I guess I'm less concerned about the details of the pipe implementation at this point.  I think something is possible there, though.  So I really think this API is right for fetch:

```
var pipe = new Pipe(0 /* buffer size */);
pipe.writer.write.then(onWritten);
var req = new Request(url, { body: pipe.reader });
```

In this case I believe its the time-disconnected Request which leads us here.  So no, I don't think its a precedent for:

```
var pipe = new Pipe(0 /* buffer size */);
startWritingMyStuff(pipe.writer);
webFS.writeToFile(filename, pipe.reader);
```

That example does not have a time-disconnected object, so it doesn't have the same problems.

To me the Request object is conceptually:

> A Request is a potential network request to be performed.  When fetch() or cache.add() consume the Request they effectively drain it to perform the actual operation.

To me this exactly mirrors a ReadableStream:

>  A ReadableStream represents a potential stream of future data.  A consumer drains it to operate on it.

I think the symmetry there makes this API just favor ReadableStream.

From a practical implementation side of things, I want this to be optimizable:

```
caches.open('mycache').then(function(cache) {
  return cache.match(oldUrl);
}).then(function(response) {
  var post = new Request(url, { body: response.body, method: 'POST' });
  return fetch(post);
}).then(function(postResponse) {
  // success
});
```

In this case response.body file data from Cache API should be streamable off-main-thread to the IPC connected to the fetch upload socket.

That seems straightforward with Request.body being a simple ReadableStream.  It seems harder to me if we have to accommodate Request.body.write() or something like it.

Finally, I will note that there are few cases where the body always being a ReadableStream even requires the pipe to be made.  You can still do things like this:

```
var req = new Request(url, {
  body: source.pipeThrough(fooTransform)
              .pipeThrough(barTransform)
              .pipeThrough(bazTransform)
});
```

Anyway, I think I have exhausted my commentary on this issue. :-)  Sorry for the wall of text.

**Comment by domenic** [comment]
> That example does not have a time-disconnected object, so it doesn't have the same problems.

OK, thanks, that does help alleviate my concerns about the precedent. It would have been nice to have a non-time-disconnected API (like Node.js has), but maybe it is not worth adding if we already have the time-disconnected version.

> I guess I'm less concerned about the details of the pipe implementation at this point. I think something is possible there, though.

I'm a little scared. But maybe we should start a whatwg/streams thread about that. I will do so.

**Comment by yutakahirano** [comment]
Sorry for being away for a long time.

So @wanderview and @domenic  agreed to pass a ReadableStream rather than passing a revealer function, relying on the pipe optimization, right? I'm not unhappy with the idea, but it looks we go back to the initial position such as [this comment] (https://github.com/yutakahirano/fetch-with-streams/issues/30#issuecomment-90084002). I just want to understand the reasoning and what we learned from the entire discussion.

- Just exposing a writable stream via a revealer function doesn't automatically provide a direct connection between a user and the final sink.
- To ensure the direct connection, we have a hard-to-use API (e.g. revelation will be delayed).
- Otherwise, we should expect pipe to be optimizable. And if we assume that, passing a ReadableStream is also optimizable and it's easier to use.

Are these correct?

**Comment by domenic** [comment]
I think so. I would quibble that it's not easier to use for some cases, and people will be forced to create pipes with some frequency, but overall those look correct.

**Comment by yutakahirano** [comment]
Thank you. @tyoshino, are you OK with the idea?

**Comment by tyoshino** [comment]
OK. Let's start with that approach.

**Comment by yutakahirano** [comment]
Thanks, I'll write something.

**Comment by annevk** [comment]
What is the latest on this? This can replace WebSocket (which is not available over H2). It would be great to get it out there.

**Comment by yutakahirano** [comment]
Sorry, no progress yet. I've concentrated on "response construction from stream" recently. I'll put more effort on this issue.

IIUC we need to address the redirect problem, right? Is there any other problems?

**Comment by annevk** [comment]
I'm not sure, but for redirects we should do the same we do today I think. Tee the request body and use the tee on a redirect (if any). If redirect mode is "error" and window is is set to null we can avoid teeing. The server can assist the user agent here by sending a response early which will allow the user agent to stop teeing.

**Comment by annevk** [comment]
The somewhat more significant issue is that streaming requires chunked encoding, I think. That is basically something browsers have not done historically and not all servers are capable of supporting. None of that is necessarily problematic, but it might require some work beyond just hooking up the API. It will certainly require some changes to Fetch which assumes a request content length in places.

**Comment by jakearchibald** [comment]
Seems that this strategy for redirects means streams aren't a good indication of upload progress. Will this need to be added as a separate feature?

In terms of chunked encoding, the chunking isn't the problem right? HTTP2 data is always delivered in frames. I guess the issue is just the lack of `content-length`.

**Comment by yutakahirano** [comment]
#24 is the issue for upload progress, and I agree that we need a separate upload progress mechanism.

**Comment by annevk** [comment]
@jakearchibald it's not any different from how progress events work today...

**Comment by annevk** [comment]
@jakearchibald the issue with chunked is that HTTP/1 is still around.

**Comment by jakearchibald** [comment]
Sure, I mean that streaming doesn't require chunked encoding if you know the content-length.

Rejecting is an option if there isn't a content-length and browser can't establish an HTTP/2 connection.

**Comment by annevk** [comment]
We cannot rely on a developer-supplied length. Not reliable. (See also past discussion.)

**Comment by annevk** [comment]
(See also earlier in this issue for this very discussion.)

**Comment by jakearchibald** [comment]
Hm, sorry. Did a search for "length" and couldn't find it. Guess I should read the whole thing.

**Comment by yutakahirano** [comment]
@annevk For content-length I created a new issue (#65).

**Comment by yutakahirano** [comment]
@annevk For chunked-encoding is this the same problem as #57?

**Comment by domenic** [comment]
Should we close this issue (since the question in the OP is settled) and use a new issue label to track uploading?

**Comment by domenic** [comment]
I guess I see the following rough outline of a solution:

- Per discussions in #57, ensure that uploading a stream never sets Content-Length and always sets TE: chunked (or maybe in HTTP/2 the actual header is not necessary?). This has an edge case summarized in https://github.com/yutakahirano/fetch-with-streams/issues/57#issuecomment-139556744 which we are OK with just warning developers about.
- Forced preflight might be required? I don't understand this issue very much. #10 
- To handle redirects, either: tee the stream (as the spec does now), or only allow supplying a body stream if the redirect mode is "error". I opened #66 to discuss.

I think we could probably write a spec for uploading now? Most of these are settled. I guess redirects is the big open question from my perspective.

**Comment by annevk** [comment]
Given the length of this issue I'm happy with someone closing it and continuing the debate in the more focused issues.

I think we're pretty close having satisfactory answers for a minimal v1.
